fix(ivy): ngcc - properly handle aliased class expressions

JoostK opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: #29078

## What is the new behavior?

Proper handling of class expressions within ngcc.

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

**WIP!**

- [x] Testcases
- [x] Commit description
- [x] Ensure `Esm2015ReflectionHost#isClass()` is aligned with `Esm2015ReflectionHost#getClassDeclaration()` (as defined in #29209): https://github.com/angular/angular/pull/29209#discussion_r265914647

JoostK committed:
fix(ivy): ngcc - properly handle aliases class expressions

In ES2015, classes could have been emitted as a variable declaration
initialized with a class expression. In certain situations, an intermediary
variable suffixed with `_1` is present such that the variable
declaration's initializer becomes a binary expression with its rhs being
the class expression, and its lhs being the identifier of the intermediate
variable. This structure was not recognized, resulting in such classes not
being considered as a class in `Esm2015ReflectionHost`.

As a consequence, the analysis of functions/methods that return a
`ModuleWithProviders` object did not take the methods of such classes into
account.

Another edge-case with such intermediate variable was that static
properties would not be considered as class members. A testcase was added
to prevent regressions.

Fixes #29078

JoostK commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1287,6 +1369,31 @@ function collectExportedDeclarations(
+ * @param declaration The declaration for which any variable declaration should be obtained.
+ * @returns the outer variable declaration if found, undefined otherwise.
+ */
+function getOuterVariableDeclarationOfDeclaration(declaration: ts.Declaration):
```
Note to self: add tests for this.

JoostK replied:
This change fixes an edge-case where classes that have the `_1` intermediate variable would not have static properties considered as member. Testcase was added to prevent regressions.

JoostK commented on packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -548,6 +548,21 @@ const MODULE_WITH_PROVIDERS_PROGRAM = [
+    name: '/src/firebase.js',
+    contents: `
+    var AngularFireModule_1;
+    let AngularFireModule = AngularFireModule_1 = class AngularFireModule {
```
Note to self: rename to generic names.

JoostK commented on packages/compiler-cli/src/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -1555,5 +1570,21 @@ describe('Fesm2015ReflectionHost', () => {
+      expect(fn.map(
+                 fn =>
+                     [fn.declaration.name !.getText(),
+                      (host.getDeclarationOfIdentifier(fn.ngModule) !.node as ts.NamedDeclaration)
```
Oh man, this is annoying. Basically, `getModuleWithProvidersFunctions` returns an `ModuleWithProvidersFunction` structure where the `ngModule` property is the identifier as used within the object literal. That identifier therefore corresponds with `MyClass_1` and not `MyClass`. So basically, this test can't verify that the right thing happens, except for calling into the host once more. This however gives zero certainty about this actually working correctly in real-life.

I'd like to hear your thoughts here @petebacondarwin.

<sup>At the very least I will format this properly</sup>

alxhub replied:
Can't we assert that the identifier which comes back is `AngularFireModule`?

JoostK replied:
@alxhub we *could* rewrite the `AngularFireModule_1` identifier to `AngularFireModule`, however that feels very dirty as we'd end up with an identifier that belongs with a class declaration instead of a node that was actually used as value in the object literal. I would very much like to know what your approach would be here, as I don't really know what makes sense.

alxhub replied:
In the `.d.ts` file, the only node which _can_ be used in the `ModuleWithProviders<T>` type is the `ts.ClassDeclaration` for `AngularFireModule`. That's what `getModuleWithProvidersFunctions` should return. The fact that `AngularFireModule_1 ` exists at all is an artifact of the way the decorated function is compiled.

JoostK replied:
I have changed `ModuleWithProvidersFunction.ngModule` from a `ts.Identifier` to the class' `Declaration`.

JoostK commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -200,16 +214,28 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+
+
+  isClass(node: ts.Node): node is ts.NamedDeclaration {
+    return super.isClass(node) || this.resolveClassDeclaration(node) !== undefined;
```
Note to self: add testscases.

JoostK commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -200,16 +214,28 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+    return super.isClass(node) || this.resolveClassDeclaration(node) !== undefined;
+  }
+
+  getDeclarationOfIdentifier(id: ts.Identifier): Declaration|null {
```
Note to self: add testscases.

JoostK commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -392,6 +418,64 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+      // If we captured an intermediate identifier such as `MyClass_1`, store its declaration
+      // for later retrieval in order to support resolving usages of `MyClass_1` to `MyClass`.
+      if (intermediateIdentifier !== null) {
+        const intermediateDeclaration = super.getDeclarationOfIdentifier(intermediateIdentifier);
```
This usage of `super` here is less than ideal (at least deserves a comment, come to think of it), however we can't call through the ES2015 host implementation as doing so would resolve the `MyClass_1` identifier to the `MyClass` declaration for subsequent registrations, whereas we need the `MyClass_1` declaration. I don't see how we can avoid subsequent registrations, as we first need to have obtained the declaration in the first place.

alxhub replied:
👍looks okay to me - do add the comment. Alternatively we could use the `ts.TypeChecker` directly.

petebacondarwin commented on packages/compiler-cli/src/ngcc/src/host/esm2015_host.ts:
```diff
@@ -49,6 +49,26 @@ export const CONSTRUCTOR_PARAMS = 'ctorParameters' as ts.__String;
+   * `analyzedSourceFilesForAliasedClasses` tracks which source files have already been processed.
+   */
+  private aliasedClassDeclarations = new Map<ts.Declaration, ts.Identifier>();
+  private analyzedSourceFilesForAliasedClasses = new Set<ts.SourceFile>();
```
I have found in the past that it is better to make these sorts of properties `protected` unless you are going to hide them behind accessors. Otherwise you find that it is awkward to override the behaviour of the methods that use them in sub-classes. The `Esm5ReflectionHost` probably doesn't need to but a `UmdReflectionHost` might well.

JoostK replied:
Personal habit, changed into `protected` :)

JoostK commented on packages/compiler-cli/ngcc/src/host/esm5_host.ts:
```diff
@@ -180,7 +180,10 @@ export class Esm5ReflectionHost extends Esm2015ReflectionHost {
   getMembersOfClass(clazz: ClassDeclaration): ClassMember[] {
-    if (super.isClass(clazz)) return super.getMembersOfClass(clazz);
+    // Do not follow ES5's resolution logic when the node resides in a .d.ts file.
+    if (clazz.getSourceFile().isDeclarationFile) {
```
@gkalpak Oh does this feel dirty. Some context:

With the ES2015 host now also calling into `getClassDeclaration` from its `isClass` implementation, the `super.isClass` call becomes equivalent to calling `this.isClass`. This incorrectly resulted in always taking this `if`. When I removed this line there was one failing testcase left:

https://github.com/angular/angular/blob/048f8de9a8fb189b61f719c891efe1f19d4de5bc/packages/compiler-cli/ngcc/test/host/esm5_host_spec.ts#L1094-L1102

Since that file is a .d.ts file, it should not follow any of ES5's custom rules. Hence, I changed the guard to look if the source file is a declaration file or not.

I am not too sure if there's other methods in the ES5 host implementation that need to special case the handling of nodes from declaration files. I myself did not realize until now that declaration nodes also pass through the ES5 host implementation.

JoostK replied:
Alternatively, instead of checking for a declaration file we could also use `!!super.getClassDeclaration(clazz)` here, although that seems odd to me.

petebacondarwin replied:
I think this will have to do for now. But we should look at changing it when we do the refactoring that calls out the "declaration/implementation/alias" forms of a class. For example, we could have an `isImplementationClass(declaration)` which for ES2015 would only return true if the `declaration` was actually a `ClassExpression` or `ClassDeclaration`...??

gkalpak replied:
The `!!super.getClassDeclaration(clazz)` check looks better to me, but I can live with the declaration file check as well.
If we decide to keep the latter, consider using the [isFromDtsFile(clazz)][1] utility instead.

[1]: https://github.com/angular/angular/blob/bf90f0d7fab6b14e6ef403fa9bca3a5b5e614631/packages/compiler-cli/src/ngtsc/util/src/typescript.ts#L23-L29

JoostK replied:
I didn't know about that util, I will use it!

JoostK replied:
Refactored `isFromDtsFile`

JoostK committed:
ci: exclude ngcc from needing g3 presubmit

In a770aa2 ngcc was moved up one folder, however the angular-robot config
was not updated to reflect this change.

JoostK committed:
fixup! fix(ivy): ngcc - properly handle aliases class expressions

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -59,12 +79,27 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+    // `var MyClass = MyClass_1 = class MyClass {};`
+    if (ts.isVariableDeclaration(node) && node.initializer !== undefined) {
+      let targetValue = node.initializer;
+      while (targetValue && isAssignment(targetValue)) {
```
Perhaps we should make `isAssignment(node: ts.Node|undefined)` then we could simplify this (and probably other uses of `isAssignment()`) code to something like:

```ts
if (ts.isVariableDeclaration(node)) {
  let targetValue = node.initializer;
  while (isAssignment(targetValue)) {
    targetValue = targetValue.right;
  }  
```

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -338,6 +412,59 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
 
+  /**
+   * Determines if the given declaration corresponds with an intermediately assigned class
+   * declaration, in which case the original identifier of the class is returned.
```
I don't think that this description is quite right.  How about:

```
Resolves a potentially aliased declaration of a class to the actual class declaration.

If the given declaration is for an intermediate variable (an alias of a class),
this function will determine an identifier to the original declaration that represents
this class.
```

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -338,6 +412,59 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * Determines if the given declaration corresponds with an intermediately assigned class
+   * declaration, in which case the original identifier of the class is returned.
+   *
+   * @param declaration The declaration to determine the original
```
The declaration to resolve.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -49,6 +49,26 @@ export const CONSTRUCTOR_PARAMS = 'ctorParameters' as ts.__String;
+   * `analyzedSourceFilesForAliasedClasses` tracks which source files have already been processed.
+   */
+  protected aliasedClassDeclarations = new Map<ts.Declaration, ts.Identifier>();
+  protected analyzedSourceFilesForAliasedClasses = new Set<ts.SourceFile>();
```
We may well use this approach for capturing other information about the source file, lazily; and also that the word `analyze` is quite overloaded already.

I propose that we use the term `preparse` or similar and then make this set more generic: `preparsedSourceFiles`.
This would also lead naturally to a method call `preparseSourceFile(sf: ts.SourceFile)` which could be a noop if the file has already been parsed.

JoostK replied:
Yeah, this deserves some rework. I understand that there's the idea of introducing `NgccSymbol` or similar that conveniently bundles all relevant info of a class, to be created only once during a more general prepare phase per source file.

I will rename this for now, let's then work on the above in follow-up work.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -338,6 +412,59 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+      return;
+    }
+
+    const intermediateIdentifier = initializer.left;
```
I think we should be careful of terminology. Sometimes we are talking about `intermediate` and sometimes `alias`. Should we choose one approach and stick with it?

JoostK replied:
Good idea. `intermediate` in this sense only applies locally to the `_1` suffixed variable, so then something like `aliased` sounds better. On the other hand however, "aliased" to me refers to a primary thing that also has a secondary identity, whereas we are dealing with the inverse relation really (we are talking about the secondary, for which we need to know the primary thing). Do we have a better name available that avoids this mismatch (or is my interpretation here too strict 😅)

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
````diff
@@ -489,8 +616,9 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
     // ```
-    if (ts.isVariableDeclaration(symbol.valueDeclaration.parent)) {
-      const variableSymbol = this.checker.getSymbolAtLocation(symbol.valueDeclaration.parent.name);
+    const outerDeclaration = getOuterVariableDeclarationOfDeclaration(symbol.valueDeclaration);
````
`outerDeclaration` => `variableDeclaration` ?

In this case (different to ESM5) there is not real "outer" declaration because there is no IIFE block to be outside.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1177,10 +1304,23 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+          `Cannot find a declaration for NgModule ${ngModuleIdentifier.text} referenced in ${declaration!.getText()}`);
+    }
+
+    return declaration && {name, ngModule, declaration, container};
```
Given that we return `null` if `declaration === null` we should just return early right back up at the top of the function when we compute `declaration`, saving all this extra work.

JoostK replied:
I noticed this too in the meantime, will update!

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1330,6 +1468,31 @@ function collectExportedDeclarations(
+ * @param declaration The declaration for which any variable declaration should be obtained.
+ * @returns the outer variable declaration if found, undefined otherwise.
+ */
+function getOuterVariableDeclarationOfDeclaration(declaration: ts.Declaration):
```
I think "outer" is not the correct term.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1330,6 +1468,31 @@ function collectExportedDeclarations(
+ */
+function getOuterVariableDeclarationOfDeclaration(declaration: ts.Declaration):
+    ts.VariableDeclaration|undefined {
+  let node = declaration.parent;
```
It is interesting that `ts.Node.parent` is not typed to be `ts.Node|undefined`. So according to the type system there is always a parent of every node??

gkalpak replied:
LoL - Apparently [typescript.d.ts][1] is lying :smiley:
[SourceFile][2] (which extends [Declaration][3], which in turn extends [Node][4]) can have `parent: undefined`, despite [Node#parent][5] being typed as `Node` :grin:

[1]: https://github.com/Microsoft/TypeScript/blob/32054f1c31b73f8c2c824e5b4fa66af481c94dfb/lib/typescript.d.ts
[2]: https://github.com/Microsoft/TypeScript/blob/32054f1c31b73f8c2c824e5b4fa66af481c94dfb/lib/typescript.d.ts#L1711-L1734
[3]: https://github.com/Microsoft/TypeScript/blob/32054f1c31b73f8c2c824e5b4fa66af481c94dfb/lib/typescript.d.ts#L539-L541
[4]: https://github.com/Microsoft/TypeScript/blob/32054f1c31b73f8c2c824e5b4fa66af481c94dfb/lib/typescript.d.ts#L485-L491
[5]: https://github.com/Microsoft/TypeScript/blob/32054f1c31b73f8c2c824e5b4fa66af481c94dfb/lib/typescript.d.ts#L490

JoostK replied:
Thanks @gkalpak, that is exactly right. TypeScript is lying in more places, for instance `Symbol.valueDeclaration` which is clearly nullable [but not defined as such](https://github.com/Microsoft/TypeScript/blob/72dc16a62753b2a6fb9f8edc98ca01c3442e501a/src/compiler/types.ts#L3687). Probably predates strict null checks.

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/ngcc_host.ts:
```diff
@@ -37,9 +37,10 @@ export interface ModuleWithProvidersFunction {
+   * refers to.
    */
-  ngModule: ts.Identifier;
+  ngModule: Declaration;
```
👍 
I wonder if we could go a step further and define this as @gkalpak's new `ClassDeclaration` type?

JoostK replied:
Ah, actually we can't. `ClassDeclaration` is a certain kind of TypeScript node, whereas `Declaration` is a higher-level interpretation of declarations for ngtsc that also tell via which module the declaration was imported from, if any.

gkalpak replied:
If we know `ngModule.node` will indeed be a `ClassDeclaration`, then we could change the definition of `Declaration` to something like:

```ts
export interface Declaration<T extends ts.Declaration = ts.Declaration> {
  node: T;
  ...
}
```

Then, the semantics of `Declaration` (without a type argument) would remain the same everywhere else, but for `ngModule` we could use `Declaration<ClassDeclaration>`.

JoostK replied:
@gkalpak I thought about that but was reluctant to proceed with that change. But I just did and it worked out quite well, see latest fixup commit.

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -1326,6 +1345,18 @@ describe('Esm2015ReflectionHost', () => {
+          program, CLASS_EXPRESSION_FILE.name, 'IntermediateAssignmentClass',
+          ts.isVariableDeclaration);
+      const intermediateIdentifier =
+          (node.initializer as ts.BinaryExpression).left as ts.Identifier;
```
It would be even better if the identifier being passed to the function under test was one found in a separate expression like `{ ngModule: IntermediateAssignmentClass_1 }`.

JoostK replied:
Very good idea, done!

petebacondarwin commented on packages/compiler-cli/ngcc/test/host/esm2015_host_spec.ts:
```diff
@@ -1537,18 +1586,23 @@ describe('Esm2015ReflectionHost', () => {
   });
 
   describe('getModuleWithProvidersFunctions()', () => {
+    function getNgModuleName(ngModule: ts.Declaration & {name?: ts.Identifier}): string {
```
Isn't this type the same as `ClassDeclaration`?

petebacondarwin replied:
Oh I see that you have gone for `name?`, which actually makes it a `ts.NamedDeclaration` I believe:

```
    interface NamedDeclaration extends Declaration {
        name?: DeclarationName;
    }
```

gkalpak replied:
The difference with `ts.NamedDelcaration` is that `name` (if it exists) is not guaranteed to be `ts.Identifier`.

petebacondarwin replied:
Oh! So it is a pretty useless type then :-)

gkalpak replied:
Exactly! (For our usecase at least - I'm sure it's useful for someone, somewhere 😛)

JoostK replied:
The `ClassDeclaration` type is indeed equivalent, will update the type on `ModuleWithProvidersFunction` as suggested and then this won't be necessary at all.

JoostK replied:
Actually, will keep as-is. We can't use `ClassDeclaration` as it has `name` being required, whereas the `ts.Declaration` stored in `ngModule.node` does not necessarily satisfy that constraint. This is worked around via using the `!` operator in this function.

petebacondarwin reviewed (changes requested):
Overall this looks very nice. 
Enough nits to request changes :-)
But you can argue against them and I'll approve.

gkalpak commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -338,6 +412,59 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+
+    const intermediateDeclaration = this.getDeclarationOfIdentifier(intermediateIdentifier);
+    if (intermediateDeclaration === null) {
+      throw new Error(`Unable to locate declaration of ${intermediateIdentifier.text}`);
```
I think it is worth printing more info about the statement (e.g. include `statement.getText()` in the error message).

JoostK commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -49,6 +49,26 @@ export const CONSTRUCTOR_PARAMS = 'ctorParameters' as ts.__String;
+  /**
+   * In ES2015, class declarations may have been down-leveled into variable declarations,
+   * initialized using a class expression. In certain scenarios, an intermediate variable
+   * is introduced to that results in code such as:
```
*so* 😅

JoostK commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -59,12 +79,27 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+    // Recognize a variable declaration of the form `var MyClass = class MyClass {}` or
+    // `var MyClass = MyClass_1 = class MyClass {};`
+    if (ts.isVariableDeclaration(node) && node.initializer !== undefined) {
+      let targetValue = node.initializer;
```
I *think* I did it like this as it clearly separates out walking the list of assignments and arriving at the right-most value, to then decide that that right-most thing is supposed to be the class (hence only then reassigning `node` after the loop). Your explanation seems just as valid though and I wouldn't mind changing it.

JoostK commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -338,6 +412,59 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * declaration, in which case the original identifier of the class is returned.
+   *
+   * @param declaration The declaration to determine the original
+   */
```
Looks like the keystrokes were sold out that day, as proven by the line above that also stops abruptly. I will try to order some more keystrokes.

JoostK committed:
fixup! fix(ivy): ngcc - properly handle aliases class expressions

JoostK committed:
fixup! fix(ivy): ngcc - properly handle aliases class expressions

alxhub commented on packages/compiler-cli/src/ngtsc/reflection/src/host.ts:
```diff
@@ -317,11 +317,11 @@ export interface Import {
  * application.
  */
-export interface Declaration {
+export interface Declaration<T extends ts.Declaration = ts.Declaration> {
```
Yay, types!

petebacondarwin commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
```diff
@@ -1206,10 +1369,10 @@ export function isAssignmentStatement(statement: ts.Statement): statement is Ass
-      expression.operatorToken.kind === ts.SyntaxKind.EqualsToken;
+export function isAssignment(node: ts.Node | undefined):
+    node is ts.AssignmentExpression<ts.EqualsToken> {
+  return !!node && ts.isBinaryExpression(node) &&
```
@alxhub what is the reasoning behind the check being in the caller?

petebacondarwin replied:
I suspect the reason for most TS checks like this do not allow `undefined` because they were written before strict null checks were implemented.

alxhub replied:
@petebacondarwin, I have a couple reasons:

1) The main one is that I prefer logic (e.g. the decision on what to do about `undefined`) be as straightforward and up-front as possible. If I read:

```javascript
if (node.expression !== undefined && isAssignment(node.expression)
```

then it's clear to me that there might not be an expression, or the expression might not be of the shape we want, and that these two cases are explicitly treated the same.

If instead I read:

```javascript
if (isAssignment(node.expression))
```

then it's not evident at all that there is a case where there isn't an expression, or that we're deciding to handle it in the same way as the case where the expression isn't correctly shaped. I might even call this function not realizing that `undefined` was a possibility for a particular mode, and I'd prefer the type system warn me about this.

In this particular circumstance, these two cases are treated the same, but that might not always be true, and I prefer when it's obvious at the call site that we're making that decision.

2) I don't believe the TS functions are like this because they predate strict null checks, because they don't actually handle `undefined` at all even at runtime.

JoostK commented on packages/compiler-cli/ngcc/src/host/esm2015_host.ts:
````diff
@@ -59,12 +82,26 @@ export class Esm2015ReflectionHost extends TypeScriptReflectionHost implements N
+   * ```
+   *
+   * Here, the intermediate `MyClass_1` assignment is optional. In the above example, the
+   * `class MyClass {}` node is returned as declaration of `MyClass`.
````
Ah, I see. That would defeat the purpose of this method as it is in fact used to resolve to the class expression, as its symbol is the one we need to deal with actually.

I think what is relevant to know here is that this method should really be `protected`, as it is only used internally. Therefore, it is of no harm that we are returning an "internal" node here. This is not marked as protected as it is used in a test.

alxhub replied:
Ah, okay. In that case:

* Let's rename it to `getInnerClassDeclaration`.
* Should the return type be `ClassDeclaration<ts.ClassDeclaration>`?
* The return type should be `|null` not `|undefined`.
* If a method is public for testing purposes, mark it as `@internal` in the jsdoc.
* Also, I note that the logic here actually doesn't have any dependency on being in the host class itself. Could it be a separate function which is more easily tested?

petebacondarwin replied:
I have been pushing for the two concepts to be "declaration" and "implementation" rather than "outer" and "inner".

In fact in Es2015 there is no block for the class expression to be "inside" of.

JoostK committed:
fixup! fix(ivy): ngcc - properly handle aliases class expressions

JoostK committed:
fixup! fix(ivy): ngcc - properly handle aliases class expressions

JoostK committed:
fixup! fix(ivy): ngcc - properly handle aliases class expressions

JoostK committed:
fixup! fix(ivy): ngcc - properly handle aliases class expressions

jasonaden commented:
Needs a rebase, then I can get this merged.

cyrilletuzi commented:
@JoostK Hi, I'm still having this issue with a lib build with latest next versions:
- Angular 8.0.0-rc.5
- CLI 8.0.0-rc.4
- ng-packagr 5.2.0

Is that normal?

petebacondarwin commented:
@cyrilletuzi you need to try the `master` builds; we have not been backporting ngcc changes to 8.0.0.
Try using `https://github.com/angular/compiler-cli-builds` as the yarn/npm target.

cyrilletuzi commented:
@petebacondarwin Thanks for your quick answer.

I'm not sure to understand: does it mean a library build with v8.0 won't be able to be compatible with Ivy?

petebacondarwin commented:
Can you confirm what you are trying to do and what errors you are getting?
If you want to compile with ivy you need to use the `master` versions when compiling.
If you are generating a library for consumption by a project compiling with ivy then it ought to be fine to compile and publish it compiled as normal (not ivy) via v8.0.0.

cyrilletuzi commented:
@petebacondarwin 

Steps to reproduce:
- `git clone https://github.com/cyrilletuzi/angular-async-local-storage`
- `cd angular-async-local-storage`
- `git checkout v8beta`
- `npm ci`
- `npm run build` (build the lib in normal mode)
- in `projects/ivy/src/app/app.module.ts`, add: 
```ts
import { StorageModule } from '@ngx-pwa/local-storage';

@NgModule({
  imports: [
    BrowserModule,
    StorageModule.forRoot({
      IDBNoWrap: true,
    })
  ]
})
export class AppModule { }
```
- `./node_modules/.bin/ng serve ivy`

Same error as in #29078:
```
ERROR in projects/ivy/src/app/app.module.ts(9,12): error TS-991010: Value at position 1 in the NgModule.importss of AppModule is not a reference: [object Object]
```

Works if I remove `StorageModule`.

JoostK commented:
@cyrilletuzi I am confused. I can't find `StorageModule` within `@ngx-pwa/local-storage` here: https://unpkg.com/@ngx-pwa/local-storage@8.0.0-beta.12/fesm5/ngx-pwa-local-storage.js

cyrilletuzi commented:
@JoostK beta.12 didn't contain the last code, the local `npm run build` will.

I've just published beta.13 if you need it.

petebacondarwin commented:
It looks like you are trying to build an example app using ivy inside your library project? Is that right?
If so then you need to use a `next` version of Angular to do the build in that `projects/ivy/src` folder. I.E. you need to specify the correct version of `@angular/compiler-cli` and friends in https://github.com/cyrilletuzi/angular-async-local-storage/blob/master/projects/ngx-pwa/local-storage/package.json

cyrilletuzi commented:
@petebacondarwin `master` branch is indeed in v7, but as mentionned in the repro steps, you need to switch to `v8beta` branch, where the last Angular v8 `next` version is used.

JoostK commented:
It appears that the CLI's integration of ngcc is unable to resolve `dist/@ngx-pwa/local-storage/package.json`. It currently relies on `require.resolve` and some quick debugging session leaves me to believe that it only searches in `node_modules` folders, so the `dist` directory spoils the fun.

Here's how the `package.json` is resolved in `@ngtools/webpack`'s integration of ngcc:

https://github.com/angular/angular-cli/blob/5df02a3de5257162c7b8a5e2c65d388debcd51f0/packages/ngtools/webpack/src/ngcc_processor.ts#L88-L105

petebacondarwin commented:
@cyrilletuzi - sorry I didn't spot that the `git checkout v8beta` was actually checking out a verison that builds with Angular `master`.

@alan-agius4 could jump in regarding the CLI integration. We do support, I believe `dist` folders but they need to be setup via the `paths` property in `tsconfig.json`, right?

JoostK commented:
@cyrilletuzi Running ngcc manually on `dist` using `node_modules/.bin/ivy-ngcc -s ./dist` lets me compile and serve the app successfully

cyrilletuzi commented:
@JoostK I confirm your debug: it works well in an external app, it's just a problem inside a local app.

When doing `ng serve` in an external app (another project with the lib installed via npm), the first time, I see the compiler doing this:
```
Compiling @angular/core : es2015 as esm2015
Compiling @angular/common : es2015 as esm2015
Compiling @angular/platform-browser : es2015 as esm2015
Compiling @angular/platform-browser-dynamic : es2015 as esm2015
Compiling @ngx-pwa/local-storage : es2015 as esm2015
```

But when doing `ng serve` in a local app (inside `dist`), I just see this (the lib is not compiled):
```
Compiling @angular/core : es2015 as esm2015
Compiling @angular/common : es2015 as esm2015
Compiling @angular/platform-browser : es2015 as esm2015
Compiling @angular/platform-browser-dynamic : es2015 as esm2015
```

cyrilletuzi commented:
@JoostK With your workaround, tests are now passing, thanks. So it's OK on my side. 👍

alan-agius4 commented:
Hi all, the `dist` folder resolution is not yet implemented in the CLI, as this was done in angular master after the version 8 branch has been created. Hence it should be available in version 8.1
